service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java

```diff
@@ -223,4 +228,29 @@ private String extractVaccinationProtocolString(Immunization.ImmunizationVaccina
 
         return vaccinationProtocol + VACCINATION_TARGET_DISEASE + targetDiseases;
     }
+
+    private String buildCode(Immunization immunization) {
+        var vaccinationProcedure = ExtensionMappingUtils.filterExtensionByUrl(immunization, VACCINATION_PROCEDURE_URL);
+        if (vaccinationProcedure.isPresent()) {
+            CodeableConcept codeableConcept = (CodeableConcept) vaccinationProcedure.get().getValue();
+            return codeableConceptCdMapper.mapCodeableConceptToCd(codeableConcept);
+        }
+        throw new EhrMapperException("Immunization vaccination procedure not present");
+    }
+
+    private String buildVaccineCode(Immunization immunization) {
+        if (immunization.hasVaccineCode() && vaccineCodeNotUNK(immunization.getVaccineCode())) {
+            var code = CodeableConceptMappingUtils.extractTextOrCoding(immunization.getVaccineCode());
+            if (code.isPresent()) {
+                return String.format(VACCINATION_CODE, code.get());
+            }
+        } else if (immunization.hasVaccineCode() && !vaccineCodeNotUNK(immunization.getVaccineCode())) {
+            return StringUtils.EMPTY;
+        }
+        throw new EhrMapperException("Immunization vaccine code not present");
+    }
+
+    private boolean vaccineCodeNotUNK(CodeableConcept codeableConcept) {
+        return  (codeableConcept.getCodingFirstRep().hasCode() && !codeableConcept.getCodingFirstRep().getCode().equals("UNK"));
```

Comment:
why is there double negation? wouldn't be enough to check id isVaccineCodeUNK?

Reply:
I think this will through npe